crates/provider/src/native.rs

```diff
@@ -2,531 +2,920 @@ use std::{
+        // remove namespace from provider
+        let inner = self.inner.write().await;
+        if let Some(provider) = inner.provider.inner.upgrade() {
+            provider.write().await.namespaces.remove(&inner.id);
+        }
 
-        let _ = self
-            .run_command(
-                vec![format!("kill -CONT {}", process.pid)],
-                NativeRunCommandOptions {
-                    is_failure_allowed: true,
-                },
-            )
-            .await?;
         Ok(())
     }
+}
 
-    // TODO: Add test
-    async fn restart(
-        &mut self,
-        node_name: &str,
-        after_secs: Option<u16>,
-    ) -> Result<bool, ProviderError> {
-        let process = self.get_process_by_node_name(node_name)?;
-
-        let _resp = self
-            .run_command(
-                vec![format!("kill -9 {:?}", process.pid)],
-                NativeRunCommandOptions {
-                    is_failure_allowed: true,
-                },
-            )
-            .await?;
-
-        // log::debug!("{:?}", &resp);
-
-        if let Some(secs) = after_secs {
-            sleep(Duration::from_secs(secs.into())).await;
-        }
+#[derive(Debug)]
+struct NativeNodeInner<FS: FileSystem + Send + Sync + Clone> {
+    name: String,
+    command: String,
+    args: Vec<String>,
+    env: Vec<(String, String)>,
+    log_path: String,
+    process: Child,
+    stdout_reading_handle: JoinHandle<()>,
+    stderr_reading_handle: JoinHandle<()>,
+    log_writing_handle: JoinHandle<()>,
+    filesystem: FS,
+    namespace: WeakNativeNamespace<FS>,
+}
 
-        let process: &mut Process =
-            self.process_map
-                .get_mut(node_name)
-                .ok_or(ProviderError::MissingNodeInfo(
-                    node_name.to_owned(),
-                    "process".into(),
-                ))?;
+impl<FS: FileSystem + Send + Sync + Clone> NativeNodeInner<FS> {}
 
-        let mapped_env: HashMap<&str, &str> = process
-            .env
-            .iter()
-            .map(|env_var| (env_var.name.as_str(), env_var.value.as_str()))
-            .collect();
+#[derive(Debug, Clone)]
+struct NativeNode<FS: FileSystem + Send + Sync + Clone> {
+    inner: Arc<RwLock<NativeNodeInner<FS>>>,
+}
 
-        let child_process: Child = Command::new(self.command.clone())
-            .arg("-c")
-            .arg(process.command.clone())
-            .envs(&mapped_env)
-            .spawn()
-            .map_err(|e| ProviderError::ErrorSpawningNode(e.to_string()))?;
+#[async_trait]
+impl<FS: FileSystem + Send + Sync + Clone + 'static> ProviderNode for NativeNode<FS> {
+    async fn name(&self) -> String {
+        self.inner.read().await.name.clone()
+    }
 
-        process.pid = child_process.id().ok_or(ProviderError::ErrorSpawningNode(
-            "Failed to get pid".to_string(),
-        ))?;
+    async fn endpoint(&self) -> Result<(IpAddr, Port), ProviderError> {
```

Comment:
Ahh this will require holding a map of the ports, at least for `rpc`/`prometheus`. If we don't want to bloat the `NodeProvider` we should store the info at `ns` level.